utilities/transactions/transaction_lock_mgr.cc

```diff
@@ -96,6 +96,68 @@ struct LockMap {
   size_t GetStripe(const std::string& key) const;
 };
 
+DeadlockInfoBuffer::DeadlockInfoBuffer(uint32_t n_latest_dlocks) {
+  paths_buffer_ = std::vector<DeadlockPath>(n_latest_dlocks);
+}
+
+void DeadlockInfoBuffer::AddNewPath(DeadlockPath path) {
+  std::lock_guard<std::mutex> lock(paths_buffer_mutex_);
+
+  if (paths_buffer_.empty()) {
+    return;
+  }
+
+  buffer_idx_ = (buffer_idx_ + 1) % paths_buffer_.size();
+  paths_buffer_[buffer_idx_] = path;
+}
+
+void DeadlockInfoBuffer::Resize(uint32_t target_size) {
+  std::lock_guard<std::mutex> lock(paths_buffer_mutex_);
+
+  uint32_t new_bufffer_idx = buffer_idx_, offset_end = target_size - 1;
+
+  if (target_size == 0 || paths_buffer_.empty()) {
+    new_bufffer_idx = 0;
+  }
+
+  else if (target_size < paths_buffer_.size()) {
+    size_t rotate_value = (buffer_idx_ + paths_buffer_.size() - offset_end) %
+                          paths_buffer_.size();
+
+    // rotate around this new first item and reset the buffer idx
+    std::rotate(paths_buffer_.begin(), paths_buffer_.begin() + rotate_value,
+                paths_buffer_.end());
+    new_bufffer_idx = offset_end;
+  }
+
+  paths_buffer_.resize(target_size);
+  buffer_idx_ = new_bufffer_idx;
+}
+
+std::vector<DeadlockPath> DeadlockInfoBuffer::PrepareBuffer() {
+  std::lock_guard<std::mutex> lock(paths_buffer_mutex_);
+
+  auto working = paths_buffer_;
+
+  if (working.empty()) {
+    return working;
+  }
+
+  uint32_t offset_begin = (buffer_idx_ + 1) % working.size();
+
+  // reverse buffer and then rotate around buffer idx offset
+  std::reverse(working.begin(), working.end());
+  std::rotate(working.begin(), working.end() - offset_begin, working.end());
```

Comment:
offset_begin could be zero, in which case you'll be calling `rotate(..., working.end(), ...)`, which probably breaks rotate.